new(libsinsp,libscap): add is_exe_writable flag (take 2)

LucaGuerra opened the pull request:
Signed-off-by: Luca Guerra <luca@guerra.sh>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area driver-kmod

/area driver-ebpf

/area libscap

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

This PR adds:
* `flags` as an argument to execve exit to be used for additional flag
* ^ the first of these, `EXE_WRITABLE` implemented
* `proc.is_exe_writable` as a field tracked by libsinsp that we can use to base Falco rules on

This flag is true if the user that executed a process can also write to the executable file that is used to spawn it or is its owner or otherwise capable. This can be meaningful in Falco rules, since in some scenarios, especially in containerized application where we follow best practices, we don't expect a non-root user to bring its own binaries and run them.

This PR was originally proposed long ago, and now it is hopefully improved because it now works with the eBPF probe as well, also considering user namespaces. The latter can be tricky to handle and there may be a better way to work with them from eBPF, in which case I'd be very happy to know.

One confusing aspect of this flag may be that while computing `is_exe_writable` the kernel already resolved symlinks, so an user-writable symlink pointing to a non-writable file will _not_ be an exe_writable, even if `proc.exepath` reports an user-writable symlink. Personally, I think this is more of an issue in how we handle `proc.exepath` and so possibly suitable for another PR.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
new(libsinsp,libscap): added proc.is_exe_writable flag
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/97#" title="Author self-approved">LucaGuerra</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **leodido** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @leodido` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["leodido"]} -->

poiana commented:
Welcome @LucaGuerra! It looks like this is your first PR to falcosecurity/libs 🎉

poiana commented:
Hi @LucaGuerra. Thanks for your PR.

I'm waiting for a [falcosecurity](https://github.com/orgs/falcosecurity/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Flibs).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

leogr reviewed (approved):
Great PR, thank you!

/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/97#pullrequestreview-799788255" title="Approved">leogr</a>*, *<a href="https://github.com/falcosecurity/libs/pull/97#" title="Author self-approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
LGTM label has been added.  <details>Git tree hash: d2d533359aaf5d4ffceddddd73e9abe3526e5733</details>

poiana commented:
@LucaGuerra: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | c109239fa1d650aaa1f683412080b9186b7ca17a | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-minimal/1459166525142339584) | true | `/test build-libs-minimal`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=97). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
New changes are detected. LGTM label has been removed.

leogr reviewed (dismissed):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 6ea0b12c6cd7d3743172b9aafe47b476f5430ac3</details>

gnosek commented on userspace/libscap/scap_procs.c:
```diff
@@ -561,6 +564,55 @@ int32_t scap_proc_fill_loginuid(scap_t *handle, struct scap_threadinfo* tinfo, c
+	int orig_uid = setfsuid(uid);
+	int orig_gid = setfsgid(gid);
+	if(setfsuid(-1) == uid && setfsgid(-1) == gid) {
+		int fd = open(proc_exe_path, O_WRONLY);
```
access/faccess(at) instead of opening the file for write?

LucaGuerra replied:
I believe I considered that but AFAICT `access` uses the real UID/GID which in case of Falco is always root. Do you know if any variant of `access` uses the fsuid/fsgid to perform the check?

LucaGuerra replied:
However, `euidaccess()` exists and can be used with `euid`/`egid`. One clear advantage of doing that instead of `open()` is that `open()` might return other errors, such as file busy.

mstemm commented on userspace/libscap/scap_procs.c:
```diff
@@ -561,6 +564,55 @@ int32_t scap_proc_fill_loginuid(scap_t *handle, struct scap_threadinfo* tinfo, c
 	}
 }
 
+int32_t scap_proc_fill_exe_writable(scap_t* handle, struct scap_threadinfo* tinfo,  uint32_t uid, uint32_t gid, const char *procdirname, const char *exetarget)
```
Don't you need to save this information to the process list as well? I would think with these changes, a process would have the exe writable flag tracked if read from proc but it would get lost when saved to a capture file.

LucaGuerra replied:
Good point. I didn't try writing a scap file and reloading it, will check.

LucaGuerra replied:
It should work now.

leogr commented:
/test build-libs-minimal

poiana commented:
New changes are detected. LGTM label has been removed.

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/libs/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [c84dac5](https://github.com/falcosecurity/libs/commits/c84dac577d2ba1c869cc632159fa4603e7cfcdc2) track is_exe_writable when collected from proc

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/97#pullrequestreview-805823791" title="Approved">leogr</a>*, *<a href="https://github.com/falcosecurity/libs/pull/97#" title="Author self-approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

leogr commented:
/retest

poiana commented:
@LucaGuerra: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-minimal/1472875377381609472) | true | `/test build-libs-minimal`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=97). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@LucaGuerra: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-minimal/1472875377381609472) | true | `/test build-libs-minimal`
build-libs-bundled-deps | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-bundled-deps/1472875377134145536) | true | `/test build-libs-bundled-deps`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=97). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@LucaGuerra: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-minimal/1472875377381609472) | true | `/test build-libs-minimal`
build-libs-bundled-deps | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-bundled-deps/1472875377134145536) | true | `/test build-libs-bundled-deps`
build-libs-with-chisels | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-with-chisels/1472875377289334784) | true | `/test build-libs-with-chisels`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=97). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@LucaGuerra: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-minimal/1472875377381609472) | true | `/test build-libs-minimal`
build-libs-bundled-deps | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-bundled-deps/1472875377134145536) | true | `/test build-libs-bundled-deps`
build-libs-with-chisels | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs-with-chisels/1472875377289334784) | true | `/test build-libs-with-chisels`
build-libs | fb3dccde827ef8bb0a9e60e2ef0a6217efe095fa | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/97/build-libs/1472875377004122112) | true | `/test build-libs`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=97). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

LucaGuerra commented:
Addressed all comments, including @ldegio 's suggestion to add a flag field for execve instead of an entire integer for a single bit so it can be used for more flags, and I think I know which one we could implement next

mstemm reviewed (approved):
The changes look right to me.

poiana commented:
LGTM label has been added.  <details>Git tree hash: 954820bfc3dffe7eaca7c9028bc8e7609f73b87b</details>

leogr reviewed (approved):
LGTM

/approve
